R/backport_linter.R

```diff
@@ -3,12 +3,17 @@
 #' Check for usage of unavailable functions. Not reliable for testing r-devel dependencies.
 #'
 #' @param r_version Minimum R version to test for compatibility
+#' @param except Character vector of functions to be excluded from linting.
+#'  Use this to list explicitly defined backports, e.g. those imported from the {backports} package or manually
```

Comment:
shouldn't we be able to reuse object_usage_linter logic to detect if either of these is true instead?

Reply:
Nope, because we can't see all files.

Reply:
Also, we'd have to parse the NAMESPACE of the linted package to check if an alternative implementation is imported, e.g. from backports.

Reply:
Are you sure you ran this in a session where the linted package wasn't loaded, backports wasn't loaded and R didn't define the function?
I expect a lint there.

Reply:
Also note that backport linter looks for SYMBOL and SYMBOL_FUNCTION_CALL with an explicit blacklist so this linter definitely has the problem.

Reply:
> and R didn't define the function?

oh whoops :facepalm: that's why it was different for work (3.6.3) vs. home (4.2.0) :)

my point is still the same -- why are we only "solving" this for `backport_linter()`, when it seems a broader issue? Isn't `codetools::checkUsage()` our friend here?

Reply:
Backport Linter doesn't look for names. Otherwise, it wouldn't properly lint e.g. when running `backport_linter("3.0.0")` under R 4.2.0.

Reply:
it's possible to work though right? since we maintain the dictionary of backports.

sorry for all the pushback, I just find this approach rather awkward for the user when it seems possible to handle this all on their behalf.

Reply:
It's impossible to work around this issue.
Imagine this setup, linting under R 4.2.0 with `backport_linter("3.0.0")`:

file1.R
```r
my_fun <- function() {
  R_user_dir("mypkg")
}
```

`object_usage_linter()` won't complain because `tools` is loaded and R 4.2.0 provides `R_user_dir`.
`backport_linter()` finds `R_user_dir` in its blacklist. How should it check if the lint should be supressed?
The function is defined (available via `tools`).
It could also be imported from `backports` via `NAMESPACE`, in which case it would be visible if the linted package was loaded first.
Or it could be defined in another file, `file2.R`, within the same package.
Or it could be conditionally imported (like we do), in which case it would come from `tools` in R 4.2.0 despite not needing a lint.

Not too keen on splitting this PR because we need the `excecpt = ` argument for our own usage.

Reply:
All of those possibilities seem knowable to me -- just a matter of effort required. I think we should set a goal to eventually eliminate this argument / populate it automatically, but am fine to leave it for another day.